mono/metadata/boehm-gc.c

```diff
@@ -585,9 +585,31 @@ push_root (gpointer key, gpointer value, gpointer user_data)
 }
 
 static void
+push_handle_stack (HandleStack* stack)
+{
+	HandleChunk *cur = stack->bottom;
+	HandleChunk *last = stack->top;
+
+	if (!cur)
+		return;
+
+	while (cur) {
+		if (cur->size > 0)
+			GC_push_all (cur->elems, (char*)(cur->elems + cur->size) + 1);
+		if (cur == last)
+			break;
+		cur = cur->next;
+
+	}
+}
+
+static void
 mono_push_other_roots (void)
 {
 	g_hash_table_foreach (roots, push_root, NULL);
+	FOREACH_THREAD (info) {
+		push_handle_stack ((HandleStack*)info->handle_stack);
```

Comment:
`handle_stack` may be NULL, I think.

In `sgen-mono.c` we skip threads that are marked as skippable or non-running or just starting up in which case the handle_stack may be NULL.  https://github.com/mono/mono/blob/5b118b84e529f7ab104e0431b274f9c5ef182a8c/mono/metadata/sgen-mono.c#L2396

Reply:
I guess I was primarily wondering about threads that are just starting up - whether boehm will ever see a thread before its handle_stack had been allocated.

Reply:
It looks like the thread info lock should always be held:
a) when attach/detaching a thread
b) during a collection

So I hope a thread is either invisible to the GC or fully initialized, but the NULL check won't hurt.